portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -24,35 +24,81 @@
 import com.google.appengine.api.datastore.DatastoreService;
 import com.google.appengine.api.datastore.DatastoreServiceFactory;
 import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
 
-/** Servlet that returns some example content. TODO: modify this file to handle comments data */
+/** Servlet that handles comment data. */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
 
-  private ArrayList<String> comments;
-
-  @Override
-  public void init() {
-    comments = new ArrayList<String>();
-  }
-
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
     String userComment = request.getParameter("text-input");
-    if (userComment == null) {
-      userComment = "";
+    if (userComment != null) {
+      storeComment(userComment);
     }
-    comments.add(userComment);
     response.sendRedirect("/index.html");
   }
 
+  /**
+    * Stores user comment as entity in datastore.
+    */
+  private void storeComment(String userComment) {
+    Entity commentEntity = new Entity("Comment");
+    commentEntity.setProperty("text", userComment);
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    datastore.put(commentEntity);
+  }
+
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    PreparedQuery results = getAllComments();
+    
+    ArrayList<String> comments = new ArrayList<String>();
+    int numberToDisplay = getNumberToDisplay(request); 
+    int numberDisplayed = 0;
+
+    for(Entity comment : results.asIterable()) {
```

Comment:
This could be more concise with Java8 streams. 

Consider using comments.subList([your params here]).forEach([your logic]).

Reply:
Sorry, I mean interpretting results.asIterable() as a stream and using those available options. But this is optional.

Reply:
other options: results.asList().subList(). There are a lot of options here :)

Reply:
I'll go look into it! It looks like it could really make my code more elegant.

Reply:
How do you feel about 
results.asIterable().forEach(comment -> 
    comments.add(new Comment((String) comment.getProperty("text"),
                                                      (String) comment.getProperty("name"),
                                                      (String) comment.getProperty("time"))));
comments = comments.subList(0, numberToDisplay - 1);

I'm having trouble converting results into a list or stream.